engine/lib/phases/phase_drop_match_guards.ml

```diff
@@ -0,0 +1,200 @@
+(* This phase removes guards from pattern matchings. It rewrites
+   them using only pattern matchings without guards.
+   See #806 and the example in tests/guards. *)
+
+open! Prelude
+
+module%inlined_contents Make (F : Features.T) = struct
+  open Ast
+  module FA = F
+
+  module FB = struct
+    include F
+    include Features.Off.Match_guard
+  end
+
+  include
+    Phase_utils.MakeBase (F) (FB)
+      (struct
+        let phase_id = Diagnostics.Phase.DropMatchGuards
+      end)
+
+  module UA = Ast_utils.Make (F)
+  module UB = Ast_utils.Make (FB)
+
+  module Implem : ImplemT.T = struct
+    let metadata = metadata
+
+    module S = struct
+      include Features.SUBTYPE.Id
+    end
+
+    [%%inline_defs dmutability]
+
+    let rec dexpr' (span : span) (expr : A.expr') : B.expr' =
+      match expr with
+      | [%inline_arms "dexpr'.*" - Match] -> auto
+      | Match { scrutinee; arms } ->
+          let arms = transform_arms (dexpr scrutinee) (List.rev arms) [] in
+          Match { scrutinee = dexpr scrutinee; arms }
+
+    and transform_arms (scrutinee : B.expr) (remaining : A.arm list)
+        (treated : B.arm list) : B.arm list =
+      match remaining with
+      | [] -> treated
+      | { arm = { arm_pat; body; guard = None }; span } :: remaining ->
+          let new_arm : B.arm =
+            {
+              arm = { arm_pat = dpat arm_pat; body = dexpr body; guard = None };
+              span;
+            }
+          in
+          transform_arms scrutinee remaining (new_arm :: treated)
+      | {
+          arm =
+            {
+              arm_pat;
+              body;
+              guard = Some { guard = IfLet { lhs; rhs; _ }; span = guard_span };
+            };
+          span;
+        }
+        :: remaining ->
+          let result_typ = dty span body.typ in
```

Comment:
Here, can you add a comment with an example in Rust syntax (you can use `<lhs>`, `<rhs>`, `<body>` et `<arm_pat>` in the example)?
I.e. an example of the matched arm + what will be the rewrite?